### PR TITLE
HS-806_809 Node scan request

### PR DESCRIPTION
**Pull request by yangli2014**
## Description

- Defined the Node scan request and response message in shared lib
- Trigger Node scan when adding new node
- Added BFF endpoint to trigger node scan by location ID

## Jira link(s)
- https://issues.opennms.org/browse/HS-806
- https://issues.opennms.org/browse/HS-809

## Checklist
* [x] Follows Horizon Stream's [development guidelines.](https://github.com/OpenNMS/horizon-stream/wiki/Development-Guidelines)
* [x] Appropriate reviewer(s) have been selected.
* [x] Jira issue(s) have been updated to "In Review".
* [x] Includes [appropriate tests.](https://github.com/OpenNMS/horizon-stream/wiki/Test-Strategy)
* [ ] Documentation has been updated as necessary.

### CODE REVIEW

File: inventory/src/main/java/org/opennms/horizon/inventory/grpc/NodeGrpcService.java [hunk 1]
```diff
@@ -85,6 +89,8 @@ public void createNode(NodeCreateDTO request, StreamObserver<NodeDTO> responseOb
             responseObserver.onCompleted();
             // Asynchronously send task sets to Minion
             executorService.execute(() -> sendTaskSetsToMinion(node));
+            scannerService.sendNodScannerTask(List.of(nodeMapper.modelToDTO(node)),
```
Comment (thomasbigger584) [inline inventory/src/main/java/org/opennms/horizon/inventory/grpc/NodeGrpcService.java:92]: typo: `sendNodeScannerTask`

File: inventory/src/main/java/org/opennms/horizon/inventory/grpc/NodeGrpcService.java [hunk 2]
```diff
@@ -85,6 +89,8 @@ public void createNode(NodeCreateDTO request, StreamObserver<NodeDTO> responseOb
             responseObserver.onCompleted();
             // Asynchronously send task sets to Minion
             executorService.execute(() -> sendTaskSetsToMinion(node));
+            scannerService.sendNodeScannerTask(List.of(nodeMapper.modelToDTO(node)),
```
Comment (cgorantla) [inline inventory/src/main/java/org/opennms/horizon/inventory/grpc/NodeGrpcService.java:92]: `scannerService.sendNodeScannerTask` can move to `sendTaskSetsToMinion`.
Eventually we will replace `taskSetService.sendDetectorTasks` with this call

File: inventory/src/main/java/org/opennms/horizon/inventory/grpc/NodeGrpcService.java [hunk 3]
```diff
@@ -172,6 +178,25 @@ public void deleteNode(Int64Value request, StreamObserver<BoolValue> responseObs
         );
     }
 
+    @Override
+    public void startNodeScanByLocationId(Int64Value request, StreamObserver<BoolValue> responseObserver) {
```
Comment (cgorantla) [inline inventory/src/main/java/org/opennms/horizon/inventory/grpc/NodeGrpcService.java:182]: Not sure if we have this specific requirement.
IMO, we will never need to do scan all nodes for a given location.

we can start with node scan on one node/ list of nodes.

File: inventory/src/main/java/org/opennms/horizon/inventory/service/taskset/ScannerTaskSetService.java [hunk 4]
```diff
@@ -89,4 +102,20 @@ private TaskDefinition addAzureScannerTask(AzureCredential credential) {
             .setSchedule(TaskUtils.DEFAULT_SCHEDULE)
             .build();
     }
+
+    private TaskDefinition createNodeScanTask(NodeDTO node) {
+        List<String> ipAddresses = node.getIpInterfacesList().stream().map(IpInterfaceDTO::getIpAddress).collect(Collectors.toList());
```
Comment (cgorantla) [inline inventory/src/main/java/org/opennms/horizon/inventory/service/taskset/ScannerTaskSetService.java:107]: for `NodeScan` task,  sending management IpInterface is enough.
Currently there is no easy way to determine if a given IpInterface is managementIp.  May be we should add a foreign id to IpInterface on Node table for management Interface.

File: shared-lib/protobuf/src/main/proto/nodescan.proto [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+  string ifName = 5;
+  int64 ifSpeed = 6;
+  int32 ifAdminStatus = 7;
+  int32 ifOperatorStatus = 8;
+  string ifAlias = 9;
+  bool isCollected = 10;
+  bool isPolled = 11;
+  int64 lastPollTime = 12;
+  int64 lastIngressTime = 13;
+  int64 lastEgressTime = 14;
```
Comment (cgorantla) [inline shared-lib/protobuf/src/main/proto/nodescan.proto:82]: can be removed

File: shared-lib/protobuf/src/main/proto/nodescan.proto [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+  string systemDesc = 3;
+  string systemLocation = 4;
+  string systemContact = 5;
+}
+
+message ServiceResult {
+  string serviceName = 1;
+  string status = 2;
+  int64 lastGoodTime = 3;
+  int64 lastFailTime = 4;
```
Comment (cgorantla) [inline shared-lib/protobuf/src/main/proto/nodescan.proto:51]: can be removed

File: inventory/src/main/java/org/opennms/horizon/inventory/service/taskset/ScannerTaskSetService.java (hunk 4 above)
Comment (yangli2014) [reply inventory/src/main/java/org/opennms/horizon/inventory/service/taskset/ScannerTaskSetService.java:107]: The node id in the IpInterface table is not enough to identify if the IP is managed or not?

File: inventory/src/main/java/org/opennms/horizon/inventory/grpc/NodeGrpcService.java (hunk 3 above)
Comment (yangli2014) [reply inventory/src/main/java/org/opennms/horizon/inventory/grpc/NodeGrpcService.java:182]: Good point I will change it to scan with one node only for now.

File: inventory/src/main/java/org/opennms/horizon/inventory/service/taskset/ScannerTaskSetService.java (hunk 4 above)
Comment (cgorantla) [reply inventory/src/main/java/org/opennms/horizon/inventory/service/taskset/ScannerTaskSetService.java:107]: For a given node, One IpAddress will be SNMP Primary on which we do SNMP Scan, that we are calling as Management IP.   Either we add a field to IpInterface to denote that as SNMP primary or add a field in node table to link to SNMP Primary interface.

**Comment by yangli2014** [comment]
@cgorantla @thomasbigger584 Can you look at this PR again?

File: inventory/src/main/java/org/opennms/horizon/inventory/service/taskset/ScannerTaskSetService.java (hunk 4 above)
Comment (yangli2014) [reply inventory/src/main/java/org/opennms/horizon/inventory/service/taskset/ScannerTaskSetService.java:105]: I prefer adding a field in the node table. In classic. It assigns an interface as SNMP primary only when the node scan detects the SNMP agent on it. Can we assume the IpInterface when we create the new node is the SNMP primary?

Comment (cgorantla) [reply inventory/src/main/java/org/opennms/horizon/inventory/service/taskset/ScannerTaskSetService.java:105]: you are right, No need to assume SNMP primary at the creation time.
We might have more interfaces on a single node in future.  We should determine SNMP primary after we get Scan results.

**Review by cgorantla — APPROVED** [review]
LGTM!

File: shared-lib/protobuf/src/main/proto/nodescan.proto [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+  string systemName = 2;
+  string systemDesc = 3;
+  string systemLocation = 4;
+  string systemContact = 5;
+}
+
+message ServiceResult {
+  string serviceName = 1;
+  string status = 2;
+  int64 lastGoodTime = 3;
```
Comment (cgorantla) [inline shared-lib/protobuf/src/main/proto/nodescan.proto:51]: can be removed

File: shared-lib/protobuf/src/main/proto/nodescan.proto [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+  int32 ifType = 4;
+  string ifName = 5;
+  int64 ifSpeed = 6;
+  int32 ifAdminStatus = 7;
+  int32 ifOperatorStatus = 8;
+  string ifAlias = 9;
+  bool isCollected = 10;
+  bool isPolled = 11;
+  int64 lastPollTime = 12;
+  int64 lastIngressTime = 13;
```
Comment (cgorantla) [inline shared-lib/protobuf/src/main/proto/nodescan.proto:79]: we can remove 76-79

**Pull request merged**